state/protocol/pebble/mutator.go

```diff
@@ -556,14 +552,15 @@ func (m *FollowerState) insert(ctx context.Context, candidate *flow.Block, certi
 			})
 		}
 
+		_, writer := tx.ReaderWriter()
 		// index the latest sealed block in this fork
-		err = transaction.WithTx(operation.IndexLatestSealAtBlock(blockID, latestSealID))(tx)
+		err = operation.IndexLatestSealAtBlock(blockID, latestSealID)(writer)
 		if err != nil {
 			return fmt.Errorf("could not index candidate seal: %w", err)
 		}
 
 		// index the child block for recovery
-		err = transaction.WithTx(procedure.IndexNewBlock(blockID, candidate.Header.ParentID))(tx)
+		err = procedure.IndexNewBlock(blockID, candidate.Header.ParentID)(tx)
```

Comment:
## ⚠️ 

The method `Extend` is called concurrently by the compliance engine (for incoming blocks from other replicas) and the node's internal hotstuff (persisting its own proposal). `Extend` calls `insert` here. The function `IndexNewBlock` is not sufficiently concurrency safe, as it modifies the list of all children of a block, which could very well cause conflicting edits, i.e. loss of information and the corruption of the protocol state.